[bug] unable to use a type alias spec of an inlined haskell function in a refined data type spec

**Problem:**
* Define and inline a predicate over a value.
* Create an alias as shorthand to express that the predicate is satisfied in a specification.
* Use the specification in a refined data type.
```haskell
{-@ inline oneFunPred @-}
oneFunPred :: Int -> Bool
oneFunPred x = x == 1

{-@ type OneTyAlias a = {v:a | oneFunPred v} @-}

{-@ data One = One { field :: OneTyAlias Int }  @-}
data One = One Int -- ERROR (ranjit said it ought to pass)
```
**Expect:**
The refined data type has a constructor with a specification requiring the predicate to be satisfied.
**Actual:**
Unbound symbol error on `oneFunPred` for the line with the data definition.
```
lib/Bug.hs:10:12: error:
    • Illegal type specification for `Bug.One`
    Bug.One :: Bug.field##Bug.One:{VV : GHC.Types.Int | Bug.oneFunPred VV} -> {VV : Bug.One | Bug.field VV == Bug.field##Bug.One
                                                                                              && Bug.field VV == Bug.field##Bug.One
                                                                                              && (is$Bug.One VV <=> true)}
    Sort Error in Refinement: {VV : int | Bug.oneFunPred VV}
    Unbound symbol Bug.oneFunPred --- perhaps you meant: Bug.field, Bug.One ?
    •
   |
10 | data One = One Int
   |            ^^^^^^^
```

---

**Workaround:**
Inline the type alias into the refined data type specification.
```haskell
{-@ inline oneFunPred @-}
oneFunPred :: Int -> Bool
oneFunPred x = x == 1

{-@ data One = One { field :: {v:Int | oneFunPred v}}  @-}
data One = One Int

-- This example passes (as expected)
```

**Version:**
Bug present in hackage releases
```
liquidhaskell 0.8.10.2
liquid-base 4.14.1.0
liquid-prelude 0.8.10.2
```
and also present in built-from-source
```
liquidhaskell@de4f003920d23a2683987c7f89db1afcf7df3b80 (Sept 16)
```

[Link to slack conv](https://liquidhaskell.slack.com/archives/C54QAL9RR/p1600451110042700)

I think this is a different issue from the type-alias one I fixed as part of #1745. If we take a look at the error:

```
      T1761.hs: FAIL (4.90s)
        tests/test.hs:503:
        Wrong exit code; output:
          > 
          > tests/pos/T1761.hs:12:12: error:
          >     • Illegal type specification for `T1761.One`
          >     T1761.One :: T1761.field##T1761.One:{VV : GHC.Types.Int | T1761.oneFunPred VV} -> {VV : T1761.One | T1761.field VV == T1761.field##T1761.One}
          >     Sort Error in Refinement: {VV : int | T1761.oneFunPred VV}
          >     Unbound symbol T1761.oneFunPred --- perhaps you meant: T1761.One, T1761.field ?
          >     • 
          >    |
          > 12 | data One = One Int
          >    |            ^^^^^^^
```

We can see how the type alias is correctly expanded, but the `oneFunPred` can't be found. I smell another name resolution problem. I'll add a bunch of traces and investigate this.

--- a/src/Language/Haskell/Liquid/Bare.hs
+++ b/src/Language/Haskell/Liquid/Bare.hs
@@ -176,12 +176,12 @@ ghcSpecEnv sp = fromListSEnv binds
   where
     emb       = gsTcEmbeds (_gsName sp)
     binds     = concat 
-                 [ [(x,        rSort t) | (x, Loc _ _ t) <- gsMeas     (_gsData sp)]
-                 , [(symbol v, rSort t) | (v, Loc _ _ t) <- gsCtors    (_gsData sp)]
-                 , [(symbol v, vSort v) | v              <- gsReflects (_gsRefl sp)]
-                 , [(x,        vSort v) | (x, v)         <- gsFreeSyms (_gsName sp), Ghc.isConLikeId v ]
-                 , [(x, RR s mempty)    | (x, s)         <- wiredSortedSyms       ]
-                 , [(x, RR s mempty)    | (x, s)         <- _gsImps sp       ]
+                 [ [(x,        rSort t) | (x, Loc _ _ t)  <- gsMeas     (_gsData sp)]
+                 , [(symbol v, rSort t) | (v, Loc _ _ t)  <- gsCtors    (_gsData sp)]
+                 , [(symbol v, vSort v) | v               <- gsReflects (_gsRefl sp)]
+                 , [(x,        vSort v) | (x, v)          <- gsFreeSyms (_gsName sp), Ghc.isConLikeId v ]
+                 , [(x, RR s mempty)    | (x, s)          <- wiredSortedSyms       ]
+                 , [(x, RR s mempty)    | (x, s)          <- _gsImps sp       ]
                  ]
     vSort     = Bare.varSortedReft emb
     rSort     = rTypeSortedReft    emb

--- a/src/Language/Haskell/Liquid/Bare/Expand.hs
+++ b/src/Language/Haskell/Liquid/Bare/Expand.hs
@@ -259,15 +259,23 @@ class Expand a where
   expand :: BareRTEnv -> F.SourcePos -> a -> a 
 
 ----------------------------------------------------------------------------------
--- | @qualifyExpand@ first qualifies names so that we can successfully resolve 
---   them during expansion. 
+-- | @qualifyExpand@ first qualifies names so that we can successfully resolve them during expansion.
+--
+-- When expanding, it's important we pass around a 'BareRTEnv' where the type aliases have been qualified as well.
+-- This is subtle, see for example T1761. In that test, we had a type alias \"OneTyAlias a = {v:a | oneFunPred v}\" where
+-- \"oneFunPred\" was marked inline. However, inlining couldn't happen because the 'BareRTEnv' had an
+-- entry for \"T1761.oneFunPred\", so the relevant expansion of \"oneFunPred\" couldn't happen. This was
+-- because the type alias entry inside 'BareRTEnv' mentioned the tuple (\"OneTyAlias\", \"{v:a | oneFunPred v}\") but
+-- the 'snd' element needed to be qualified as well, before trying to expand anything.
 ----------------------------------------------------------------------------------
-qualifyExpand :: (Expand a, Bare.Qualify a) 
-              => Bare.Env -> ModName -> BareRTEnv -> F.SourcePos -> [F.Symbol] -> a -> a 
+qualifyExpand :: (PPrint a, Expand a, Bare.Qualify a)
+              => Bare.Env -> ModName -> BareRTEnv -> F.SourcePos -> [F.Symbol] -> a -> a
 ----------------------------------------------------------------------------------
 qualifyExpand env name rtEnv l bs
-  = expand rtEnv l
-  . Bare.qualify env name l bs
+  = expand qualifiedRTEnv l . Bare.qualify env name l bs
+  where
+    qualifiedRTEnv :: BareRTEnv
+    qualifiedRTEnv = rtEnv { typeAliases = M.map (Bare.qualify env name l bs) (typeAliases rtEnv) }
 
 ----------------------------------------------------------------------------------
 expandLoc :: (Expand a) => BareRTEnv -> Located a -> Located a 
@@ -277,7 +285,7 @@ instance Expand Expr where
   expand = expandExpr 
 
 instance Expand F.Reft where
-  expand rtEnv l (F.Reft (v, ra)) = F.Reft (v, expand rtEnv l ra) 
+  expand rtEnv l (F.Reft (v, ra)) = F.Reft (v, expand rtEnv l ra)
 
 instance Expand RReft where
   expand rtEnv l = fmap (expand rtEnv l)
@@ -298,9 +306,9 @@ instance Expand SpecType where
 
 -- | @expand@ on a BareType actually applies the type- and expression- aliases.
 instance Expand BareType where 
-  expand rtEnv l 
-    = expandReft     rtEnv l -- apply expression aliases 
-    . expandBareType rtEnv l -- apply type       aliases 
+  expand rtEnv l
+    = expandReft     rtEnv l -- apply expression aliases
+    . expandBareType rtEnv l -- apply type       aliases
 
 instance Expand (RTAlias F.Symbol Expr) where 
   expand rtEnv l x = x { rtBody = expand rtEnv l (rtBody x) } 
@@ -339,7 +347,7 @@ instance Expand BareDef where
     , body  = expand rtEnv l (body  d) 
     } 
 
-instance Expand Ms.BareSpec where 
+instance Expand Ms.BareSpec where
   expand = expandBareSpec
 
 instance Expand a => Expand (F.Located a) where 
@@ -357,9 +365,10 @@ instance Expand a => Expand [a] where
 instance Expand a => Expand (M.HashMap k a) where 
   expand rtEnv l = fmap (expand rtEnv l) 
 
+-- | Expands a 'BareSpec'.
 expandBareSpec :: BareRTEnv -> F.SourcePos -> Ms.BareSpec -> Ms.BareSpec
-expandBareSpec rtEnv l sp = sp 
-  { measures   = expand rtEnv l (measures   sp) 
+expandBareSpec rtEnv l sp = sp
+  { measures   = expand rtEnv l (measures   sp)
   , asmSigs    = expand rtEnv l (asmSigs    sp)
   , sigs       = expand rtEnv l (sigs       sp)
   , localSigs  = expand rtEnv l (localSigs  sp)
@@ -367,9 +376,9 @@ expandBareSpec rtEnv l sp = sp
   , ialiases   = [ (f x, f y) | (x, y) <- ialiases sp ]
   , dataDecls  = expand rtEnv l (dataDecls  sp)
   , newtyDecls = expand rtEnv l (newtyDecls sp)
-  } 
-  where f      = expand rtEnv l 
-  
+  }
+  where f      = expand rtEnv l
+
 expandBareType :: BareRTEnv -> F.SourcePos -> BareType -> BareType 
 expandBareType rtEnv _ = go 
   where
@@ -632,8 +641,8 @@ expandSym rtEnv l s' = expandEApp rtEnv l (EVar s', [])
 
 expandEApp :: BareRTEnv -> F.SourcePos -> (Expr, [Expr]) -> Expr
 expandEApp rtEnv l (EVar f, es) = case mBody of
-    Just re -> expandApp l   re       es' 
-    Nothing -> F.eApps       (EVar f) es' 
+    Just re -> expandApp l   re       es'
+    Nothing -> F.eApps       (EVar f) es'
   where
     eAs     = exprAliases rtEnv
     mBody   = Misc.firstMaybes [M.lookup f eAs, M.lookup (GM.dropModuleUnique f) eAs]

--- a/src/Language/Haskell/Liquid/Bare/Resolve.hs
+++ b/src/Language/Haskell/Liquid/Bare/Resolve.hs
@@ -398,6 +398,14 @@ qualifyBareSpec env name l bs sp = sp
   } 
   where f      = qualify env name l bs 
 
+instance Qualify a => Qualify (RTAlias F.Symbol a) where
+  qualify env name l bs rtAlias
+   = rtAlias { rtName  = qualify env name l bs (rtName rtAlias)
+             , rtTArgs = qualify env name l bs (rtTArgs rtAlias)
+             , rtVArgs = qualify env name l bs (rtVArgs rtAlias)
+             , rtBody  = qualify env name l bs (rtBody rtAlias)
+             }
+
 instance Qualify F.Expr where 
   qualify = substEnv 
 

--- a/tests/pos/T1761.hs
+++ b/tests/pos/T1761.hs
@@ -1,0 +1,11 @@
+
+module T1761 where
+
+{-@ inline oneFunPred @-}
+oneFunPred :: Int -> Bool
+oneFunPred x = x == 1
+
+{-@ type OneTyAlias a = {v:a | oneFunPred v} @-}
+
+{-@ data One = One { field :: OneTyAlias Int }  @-}
+data One = One Int

